add NATS support

murali-reddy opened the pull request:
**Issue Ref**: #669 
 
**Description**: 

Add NATS support into Kubeless.

[PR Description]

This PR adds a first-cut integration of NATS into Kubeless. NATS provides wide variety of messaging options, this PR adds suport for https://nats.io/documentation/concepts/nats-queueing/

Please see the documentation included in the PR on how to use the functionality.

To the reviewer:

Commits are logically organised to make easier to follow.
```
8a38f8c6cb546cdbac8b3899f96341adc3e02fb9 (HEAD -> nats, origin/nats) NATS documentation
0fdb77e3f517c45fe5c74c978cbaa9fbb3bec9b7 fix tests
085e38f501126216c941ce3fb18194d0487444d7 build, CI and other misc stuff
0868bc84656ec1ff58955e5f3514609561232303 CLI support for NATS
90427b67f07810b9cc6df43c510bf11e3648c25d core nats trigger functionality
2254359450820910968bdd54646ffae72fa4f563 auto-generated code from code generator
b904a1d33d1fe325becfb8db380fee85ac6f607b API spec for nats trigger
2345b2a614c304adffd927aede296ad5caaf8e7a vendor nats-io/go-nats and nats-io/nuid
```

**TODOs**:
 - [x] Ready to review
 - [x] Automated Tests
 - [x] Docs

Planning to add some more IT tests, but PR is good to get started with review.

wallyqs commented on pkg/event-consumers/nats/nats-consumer.go:
```diff
@@ -0,0 +1,178 @@
+	if err != nil {
+		return err
+	}
+	if res.StatusCode != 200 {
```
should do something like `defer resp.Body.Close()` here to close response body and avoid leaking file descriptors ;)

andresmgot commented on cmd/kubeless/trigger/nats/publish.go:
```diff
@@ -0,0 +1,76 @@
+	},
+}
+
+func publishTopic(topic, message, url string) error {
```
nice solution :)

andresmgot commented on .travis.yml:
```diff
@@ -139,6 +154,11 @@ script:
+        make nats-controller-image NATS_CONTROLLER_IMAGE=$NATS_CONTROLLER_IMAGE
+        ./script/integration-tests minikube deployment
+        ./script/integration-tests minikube nats
+        ;;
```
the image should be done in the `GKE` scenario as well since it is the scenario that pushes the images and test everything

andresmgot replied:
also retag the image with the TRAVIS_TAG in the `after_success` hook, rename the file in the `before_deploy` and finally add the file in the `deploy.file` list

murali-reddy replied:
Done.

andresmgot commented on script/integration-tests:
```diff
@@ -75,6 +75,9 @@ basic)
     bats tests/integration-tests-kafka.bats
     ;;
+nats)
+    bats tests/integration-tests-nats.bats
```
add this tests as well in the `*` case (for GKE)

murali-reddy replied:
I will do it in a seperate PR. Intergration tests for NATS are running as part of Minikube which should be sufficient for now.

andresmgot commented on script/nats-controller.sh:
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
```
you should be able to reuse the `binary-controller` script instead of adding a new one

andresmgot commented on docs/quick-start.md:
```diff
@@ -133,7 +133,11 @@ Kubeless also supports [ingress](https://kubernetes.io/docs/concepts/services-ne
 ## PubSub function
 
-We provide several [PubSub runtimes](https://hub.docker.com/r/kubeless/), which has suffix `event-consumer`, which help you to quickly deploy your function with PubSub mechanism. The PubSub function will expect to consume input messages from a predefined Kafka topic which means Kafka is required. In Kubeless [release page](https://github.com/kubeless/kubeless/releases), you can find the manifest  ...
+We provide several [PubSub runtimes](https://hub.docker.com/r/kubeless/), which has suffix `event-consumer`, which help you to quickly deploy your function with PubSub mechanism. The PubSub function will expect to consume input messages from a predefined topic from a messaging system. Kubeless supports using events from Kafka and NATS messaging systems as Triggers.
```
> We provide several [PubSub runtimes](https://hub.docker.com/r/kubeless/), which has suffix `event-consumer`, which help you to quickly deploy your function with PubSub mechanism

that is no longer correct. I would say something like "all the available runtimes can be triggered with a PubSub mechanism..."

andresmgot commented on docs/quick-start.md:
```diff
@@ -190,6 +194,59 @@ $ kubectl logs test-695251588-cxwmc
+
+Above command will create NATS cluster IP service `nats.nats-io.svc.cluster.local:4222` which is the default URL Kubeless NATS trigger contoller expects.
+
+At this point you are all set try Kubeless NATS triggers.
```
I would continue the tutorial a little bit more and deploy a function using a NATS topic

murali-reddy replied:
Expanded the quick start guide for NATS.

andresmgot commented:
We should also add an additional flag to the `kubeless function deploy ` to complement the `--trigger-topic` flag and knows if the target is Kafka or NATS. Or just remove the `--trigger-topic` flag to clarify that PubSub triggers have a different life cycle and should be created/deleted independently. I would do the second since right know if you create and delete a function with `--trigger-topic` an orphaned KafkaTrigger is left.

wallyqs commented on pkg/event-consumers/nats/nats-consumer.go:
```diff
@@ -0,0 +1,179 @@
+var (
+	stopM     map[string](chan struct{})
+	stoppedM  map[string](chan struct{})
+	consumerM map[string]bool
```
Just wondering but is this map ok in terms of concurrent access? It seems would be used when either `syncNATSTrigger` and `FunctionAddedDeletedUpdated` are called but not sure if they would both happen concurrently.

murali-reddy replied:
Added locking there is a remote chance of race condition.

murali-reddy commented:
@andresmgot I have removed `--trigger-topic` flag in the `kubeless function deploy` call. Updated quick start guide for `Kafka` as well to reflect this.

murali-reddy commented:
thanks for the review @andresmgot @wallyqs 

@wallyqs you should be able to use the functionality by building image from the master. Ping me on slack if you run into any issue. Would be nice to get your feedback and possible future enhacements.
